src/ast/irbuilderbpf.cpp

```diff
@@ -533,6 +538,32 @@ Value *IRBuilderBPF::CreateWriteMapValueAllocation(const SizedType &value_type,
       loc);
 }
 
+Value *IRBuilderBPF::CreateVariableAllocationInit(const SizedType &value_type,
+                                                  const std::string &name,
+                                                  const location &loc)
+{
+  /* Variables are lexically scoped but the same variable name may reuse
```

Comment:
Does this comment still apply? e.g. the two variables below are not the same (as they have a different scope, as per your logic below). Also, just to be clear, the hoisting doesn't behavior right because the slots/ids are different for these two correct?

Reply:
So this comment does apply because the variables do use the same slot:
```
if (arg0 > 0) { $x = 1 } else { $x = 2 }
```

But...these variables do NOT use the same slot:
```
@[1] = 1; for ($kv : @) { $y = $kv } for ($kv : @) { $y = $kv } 
```

Its basically because when using for loops, we do a bunch of logic to add/remove variables from the current scope while for if/else, we do not alter the `variables_` map.

But to my point above, I can't find a repro where not hoisting would cause bugs. So maybe I'll just remove it.

Reply:
`if (arg0 > 0) { $x = 1 } else { $x = 2 }` - but these two variable have different entries in `variables_` right because they have different scopes or am I missing some implementation detail?

Reply:
@ryantimwilson Ah, sorry, I was looking at `variables_` in semantic analyser. That being said, it seems we still have an issue with using only the variable names, which is broken in trunk - probably something I missed with the addition of block scoping.

Here is an example of the issue:
```
BEGIN {
	if (1) {
		$x = 1;
	}
	
	if (1) { 
		let $x; 
		if (0) {
			$x = 5;
		}
		print($x)
	} 
	exit(); 
}
```

This prints 1, when it should print 0.

Let me know if you want to try to fix forward with this PR or if I should put up one first that fixes this (as it's probably touching the same code).

Reply:
@jordalgo go ahead and fix this in another PR, would appreciate it! Gives me time to debug some other stuff

Reply:
This is almost ready: https://github.com/bpftrace/bpftrace/pull/3550

Reply:
Looks like your PR is merged, I will update mine